File: src/Scoreboard.Domain/TopScoreboard.cs
namespace Scoreboard.Domain;

public sealed class TopScoreboard : IScoreboard
{
    private const int MaxSupportedCapacity = 10;

    private readonly Dictionary<string, ScoreNode> _nodesByPlayer = new(StringComparer.OrdinalIgnoreCase);

    private ScoreNode? _head;
    private ScoreNode? _tail;
    private int _count;

    public TopScoreboard(int capacity = MaxSupportedCapacity)
    {
        if (capacity <= 0 || capacity > MaxSupportedCapacity)
        {
            throw new ArgumentOutOfRangeException(
                nameof(capacity),
                $"Capacity must be between 1 and {MaxSupportedCapacity}.");
        }

        Capacity = capacity;
    }

    public int Capacity { get; }

    public int Count => _count;

    public bool UpsertScore(string playerName, int score)
    {
        var candidate = new PlayerScore(playerName, score);

        if (_nodesByPlayer.TryGetValue(candidate.PlayerName, out ScoreNode? existingNode))
        {
            RemoveNode(existingNode);
            _nodesByPlayer.Remove(candidate.PlayerName);
            _count--;
        }

        if (_count == Capacity && _tail is not null && CompareByRank(candidate, _tail.Value) > 0)
        {
            return false;
        }

        var newNode = new ScoreNode(candidate);
        InsertNodeByRank(newNode);
        _nodesByPlayer[candidate.PlayerName] = newNode;
        _count++;

        if (_count > Capacity)
        {
            TrimTail();
        }

        return _nodesByPlayer.ContainsKey(candidate.PlayerName);
    }

    public bool Remove(string playerName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(playerName);

        var key = playerName.Trim();

        if (!_nodesByPlayer.TryGetValue(key, out ScoreNode? node))
        {
            return false;
        }

        RemoveNode(node);
        _nodesByPlayer.Remove(key);
        _count--;
        return true;
    }

    public IReadOnlyList<PlayerScore> GetScores()
    {
        var results = new List<PlayerScore>(_count);
        ScoreNode? current = _head;

        while (current is not null)
        {
            results.Add(current.Value);
            current = current.Next;
        }

        return results;
    }

    private static int CompareByRank(in PlayerScore left, in PlayerScore right)
    {
        int scoreComparison = right.Score.CompareTo(left.Score);
        if (scoreComparison != 0)
        {
            return scoreComparison;
        }

        return StringComparer.OrdinalIgnoreCase.Compare(left.PlayerName, right.PlayerName);
    }

    private void InsertNodeByRank(ScoreNode node)
    {
        if (_head is null)
        {
            _head = node;
            _tail = node;
            return;
        }

        ScoreNode? current = _head;
        while (current is not null && CompareByRank(node.Value, current.Value) >= 0)
        {
            current = current.Next;
        }

        if (current is null)
        {
            AppendToTail(node);
            return;
        }

        InsertBefore(current, node);
    }

    private void AppendToTail(ScoreNode node)
    {
        if (_tail is null)
        {
            _head = node;
            _tail = node;
            return;
        }

        node.Previous = _tail;
        _tail.Next = node;
        _tail = node;
    }

    private void InsertBefore(ScoreNode target, ScoreNode node)
    {
        node.Next = target;
        node.Previous = target.Previous;

        if (target.Previous is null)
        {
            _head = node;
        }
        else
        {
            target.Previous.Next = node;
        }

        target.Previous = node;
    }

    private void RemoveNode(ScoreNode node)
    {
        if (node.Previous is null)
        {
            _head = node.Next;
        }
        else
        {
            node.Previous.Next = node.Next;
        }

        if (node.Next is null)
        {
            _tail = node.Previous;
        }
        else
        {
            node.Next.Previous = node.Previous;
        }

        node.Next = null;
        node.Previous = null;
    }

    private void TrimTail()
    {
        if (_tail is null)
        {
            return;
        }

        string playerName = _tail.Value.PlayerName;
        RemoveNode(_tail);
        _nodesByPlayer.Remove(playerName);
        _count--;
    }

    private sealed class ScoreNode
    {
        public ScoreNode(PlayerScore value)
        {
            Value = value;
        }

        public PlayerScore Value { get; }

        public ScoreNode? Previous { get; set; }

        public ScoreNode? Next { get; set; }
    }
}

File: src/com/scoreboard/domain/TopScoreboard.java
package com.scoreboard.domain;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class TopScoreboard implements Scoreboard {

  private static final int MAX_SUPPORTED_CAPACITY = 10;

  private final int capacity;
  private final Map<String, ScoreNode> nodesByPlayer = new HashMap<>();

  private ScoreNode head;
  private ScoreNode tail;
  private int count;

  public TopScoreboard() {
    this(MAX_SUPPORTED_CAPACITY);
  }

  public TopScoreboard(int capacity) {
    if (capacity <= 0 || capacity > MAX_SUPPORTED_CAPACITY) {
      throw new IllegalArgumentException(
          "Capacity must be between 1 and " + MAX_SUPPORTED_CAPACITY + ".");
    }

    this.capacity = capacity;
  }

  @Override
  public int capacity() {
    return capacity;
  }

  @Override
  public int count() {
    return count;
  }

  @Override
  public boolean upsertScore(String playerName, int score) {
    PlayerScore candidate = new PlayerScore(playerName, score);
    String normalizedKey = normalizeKey(candidate.playerName());

    ScoreNode existingNode = nodesByPlayer.get(normalizedKey);
    if (existingNode != null) {
      removeNode(existingNode);
      nodesByPlayer.remove(normalizedKey);
      count--;
    }

    if (count == capacity && tail != null && compareByRank(candidate, tail.value) > 0) {
      return false;
    }

    ScoreNode newNode = new ScoreNode(candidate);
    insertNodeByRank(newNode);
    nodesByPlayer.put(normalizedKey, newNode);
    count++;

    if (count > capacity) {
      trimTail();
    }

    return nodesByPlayer.containsKey(normalizedKey);
  }

  @Override
  public boolean remove(String playerName) {
    if (playerName == null || playerName.isBlank()) {
      throw new IllegalArgumentException("Player name must not be blank.");
    }

    String normalizedKey = normalizeKey(playerName);
    ScoreNode node = nodesByPlayer.get(normalizedKey);
    if (node == null) {
      return false;
    }

    removeNode(node);
    nodesByPlayer.remove(normalizedKey);
    count--;
    return true;
  }

  @Override
  public List<PlayerScore> getScores() {
    List<PlayerScore> scores = new ArrayList<>(count);
    ScoreNode current = head;

    while (current != null) {
      scores.add(current.value);
      current = current.next;
    }

    return List.copyOf(scores);
  }

  private static String normalizeKey(String playerName) {
    return playerName.trim().toLowerCase(Locale.ROOT);
  }

  private static int compareByRank(PlayerScore left, PlayerScore right) {
    int scoreComparison = Integer.compare(right.score(), left.score());
    if (scoreComparison != 0) {
      return scoreComparison;
    }

    int caseInsensitiveNameComparison = left.playerName().compareToIgnoreCase(right.playerName());
    if (caseInsensitiveNameComparison != 0) {
      return caseInsensitiveNameComparison;
    }

    return left.playerName().compareTo(right.playerName());
  }

  private void insertNodeByRank(ScoreNode node) {
    if (head == null) {
      head = node;
      tail = node;
      return;
    }

    ScoreNode current = head;
    while (current != null && compareByRank(node.value, current.value) >= 0) {
      current = current.next;
    }

    if (current == null) {
      appendToTail(node);
      return;
    }

    insertBefore(current, node);
  }

  private void appendToTail(ScoreNode node) {
    if (tail == null) {
      head = node;
      tail = node;
      return;
    }

    node.previous = tail;
    tail.next = node;
    tail = node;
  }

  private void insertBefore(ScoreNode target, ScoreNode node) {
    node.next = target;
    node.previous = target.previous;

    if (target.previous == null) {
      head = node;
    } else {
      target.previous.next = node;
    }

    target.previous = node;
  }

  private void removeNode(ScoreNode node) {
    if (node.previous == null) {
      head = node.next;
    } else {
      node.previous.next = node.next;
    }

    if (node.next == null) {
      tail = node.previous;
    } else {
      node.next.previous = node.previous;
    }

    node.next = null;
    node.previous = null;
  }

  private void trimTail() {
    if (tail == null) {
      return;
    }

    String normalizedKey = normalizeKey(tail.value.playerName());
    removeNode(tail);
    nodesByPlayer.remove(normalizedKey);
    count--;
  }

  private static final class ScoreNode {

    private final PlayerScore value;
    private ScoreNode previous;
    private ScoreNode next;

    private ScoreNode(PlayerScore value) {
      this.value = value;
    }
  }
}

